File: Java/leetcode75/src/main/java/org/phoenix/leetcode/challenges/Problem41_SearchInABinarySearchTree.java
package org.phoenix.leetcode.challenges;

import org.phoenix.leetcode.challenges.Problem33_MaximumDepthOfBinaryTree.TreeNode;

import java.util.LinkedList;
import java.util.Queue;

public class Problem41_SearchInABinarySearchTree {
    public TreeNode searchBST(TreeNode root, int val) {
        if (root == null) return null;
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        while (!queue.isEmpty()) {
            int size = queue.size();
            while (size > 0) {
                TreeNode curr = queue.poll();
                if (curr.val == val) return curr;
                if (curr.right != null) queue.offer(curr.right);
                if (curr.left != null) queue.offer(curr.left);
                size--;
            }
        }
        return null;
    }
}

File: JavaScript/leetcode75/src/main/Problem41_SearchInABinarySearchTree.js
/**
 * @param {TreeNode} root
 * @param {number} val
 * @return {TreeNode}
 */
var searchBST = function (root, val) {
  if (root == null || root.val == val) return root;
  if (root.val > val) return searchBST(root.left, val);
  return searchBST(root.right, val);
};

module.exports = searchBST;
